### PR TITLE
Teacher Registration Flow

### LINKED ISSUE #409: Teacher Account Join Flow
Need:

- [x] Final Copy https://docs.google.com/spreadsheets/d/1IlqxvbaKPqCbRZaUOqWPO0LyDbaqBqQuJDDCq5kwLqg/edit (ping me if you need access) 
- [x] Final Designs (@carljbowman)

### PR DESCRIPTION
**Pull request by rschamp**
WIP
Resolves #409, requires LLK/scratchr2#3728

* Add a view for registering teacher accounts
* Update form components sitewide to use `formsy-react`, `formsy-react-components`
* Remove all mixins
* Remove the last global `API_HOST`, use webpack's `DefinePlugin` instead
* Update `lib/api` to allow passing `GET` params and to submit `POST` form data with urlencoded parameters

### CODE REVIEW

File: src/components/deck/deck.scss [hunk 1]
```diff
@@ -0,0 +1,7 @@
+@import "../../frameless";
+
+@include responsive-layout (".deck", ".slide");
+
+.deck {
+    min-height: 100vh;
```
Comment (thisandagain) [inline src/components/deck/deck.scss]: Does our webpack config handle fallbacks for this when not supported?
http://caniuse.com/#feat=viewport-units

File: src/components/forms/form.jsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        };
+    },
+    render: function () {
+        var classes = classNames(
+            'form',
+            this.props.className
+        );
+        return (
+          <Formsy.Form {... this.props} className={classes}>
+            <GeneralError name="all" />
```
Comment (thisandagain) [inline src/components/forms/form.jsx]: Inconsistent indentation.

File: src/components/forms/input-hoc.jsx [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        getDefaultProps: function () {
+            return {
+                messages: {
+                    'general.notRequired': 'Not Required'
+                }
+            };
+        },
+        render: function () {
+            return (
+                <Component help={this.props.required ? null : this.props.messages['general.notRequired']}
```
Comment (thisandagain) [inline src/components/forms/input-hoc.jsx]: Might be easier to read if we move the ternary into a var at the top of the `render` function.

File: src/components/forms/input.jsx [hunk 4]
```diff
@@ -1,22 +1,27 @@
...
+var inputHOC = require('./input-hoc.jsx');
 
 require('./input.scss');
+require('./row.scss');
 
 var Input = React.createClass({
     type: 'Input',
-    propTypes: {
-        
+    getDefaultProps: function () {
```
Comment (thisandagain) [inline src/components/forms/input.jsx]: Is this functionally equivalent to simply removing the `getDefaultProps` attribute?

File: src/components/forms/phone-input.jsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    render: function () {
+        var defaultCountry = PhoneInput.getDefaultProps().defaultCountry;
+        if (allIso2.indexOf(this.props.defaultCountry.toLowerCase()) !== -1) {
+            defaultCountry =  this.props.defaultCountry.toLowerCase();
+        }
+        return (
+            <Row {... this.getRowProperties()}
+                 htmlFor={this.getId()}
+                 className={classNames('phone-input', this.props.className)}
+            >
```
Comment (thisandagain) [inline src/components/forms/phone-input.jsx]: Is our style to do this or:

```js
<Some foo={foo}
    bar={bar}
    baz={baz} />
```

File: src/components/languagechooser/languagechooser.jsx [hunk 6]
```diff
@@ -1,46 +1,42 @@
...
+var Form = require('../forms/form.jsx');
 var Select = require('../forms/select.jsx');
 
 /**
  * Footer dropdown menu that allows one to change their language.
  */
 var LanguageChooser = React.createClass({
     type: 'LanguageChooser',
-    mixins: [
-        Api
```
Comment (thisandagain) [inline src/components/languagechooser/languagechooser.jsx]: 🙌

File: src/components/languagechooser/languagechooser.jsx [hunk 7]
```diff
@@ -1,46 +1,42 @@
...
-                                </option>;
-                    }.bind(this))}
-                </Select>
-            </div>
+            <Form className={classes}>
+                <Select name="language"
+                        options={languageOptions}
+                        defaultValue={this.props.locale}
+                        onChange={this.onSetLanguage}
+                        required />
```
Comment (thisandagain) [inline src/components/languagechooser/languagechooser.jsx]: See other note about inconsistent styles for terminating an indented multi-line tag.

File: src/components/slide/slide.scss [hunk 8]
```diff
@@ -0,0 +1,2 @@
+.slide {
```
Comment (thisandagain) [inline src/components/slide/slide.scss]: Stub?

File: src/lib/country-data.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+module.exports = {};
+var countries = module.exports.data = require('iso-3166-2').data;
+
+module.exports.countryOptions = Object.keys(countries).map(function (code) {
+    return {value: code.toLowerCase(), label: countries[code].name};
+}).sort(function (a, b) {
+    return a.label < b.label ? -1 : 1;
+});
+
+module.exports.subdivisionOptions =
```
Comment (thisandagain) [inline src/lib/country-data.js]: Awkward line break.

File: src/views/teacherregistration/teacherregistration.jsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+    },
+    render: function () {
+        return (
+            <Deck className="teacher-registration">
+                {this.state.registrationError ?
+                    <Steps.RegistrationError {... this.state} />
+                :
+                    <Progression {... this.state}>
+                        <Steps.UsernameStep     onNextStep={this.advanceStep}
```
Comment (thisandagain) [inline src/views/teacherregistration/teacherregistration.jsx]: Another style inconsistency here. Curious to hear if you have any thoughts on how to unify this.

**Comment by thisandagain** [comment]
This looks great. `src/components/registration/steps.jsx` is *huge*, but makes sense and seems well organized given the amount of functionality it is orchestrating. Mostly just questions and some minor nitpicks left in line comments.

File: src/components/slide/slide.scss (hunk 8 above)
Comment (mewtaylor) [inline src/components/slide/slide.scss]: I'll be making a bit more use of this once I push up the formatting

File: src/components/forms/phone-input.jsx (hunk 5 above)
Comment (mewtaylor) [inline src/components/forms/phone-input.jsx]: My understanding was that we were doing the latter

**Pull request merged**